### ISSUE TITLE
PPO2 + VecNormalize 

### ISSUE THREAD

**Issue by andytwigg** [openai/baselines#630]
@pzhokhov Is there any 'best practise' on using PPO2+VecNormalize together? Since PPO2 internally normalizes advantages, I'm wondering if using VecNormalize adds another layer of dynamic scaling that may make things harder to learn. On the other hand, if it only scales advantages, it means that the value network is still required to output returns on the same scale as it receives, which may unrealistic (eg humanoid). It would be interesting to know if there's any general guidance on this...

**Comment by pzhokhov** [comment]
Deferring to @joschu and @unixpickle for more comprehensive view on why are we using VecNormalize. 
Note that advantages in PPO2 are normalized only within a batch, whereas VecNormalize computes normalization based on the entire experience of the environment; thus VecNormalize normalization changes on a time scale much slower than that of advantage normalization in PPO2; so I don't think these normalizations interfere. 
Somewhat related question here: https://github.com/openai/baselines/issues/629

**Comment by andytwigg** [comment]
Yes, that makes sense (and after the discussion in #538 about fixing
vecnormalize), but I wondered if there was general guidance as I've been
having mixed results using VecNormalize.

**Comment by doviettung96** [comment]
Hi guys,
The PPO paper suggests that we should used VecNormalize to env such as Mujoco Humanoid or Roboschool Humanoid. I am wondering if I should use VecNormalize for both observation and reward or just the observation? In the case we normalizes the reward, the Tensorboard graph of the mean reward will have small values (like 10 compare to 1000 of not normalizing reward). Maybe in that case, we should modify the code to draw graph of the original reward. Any suggestion?

**Comment by pzhokhov** [comment]
Hi @doviettung96! 
Even if you are using reward-modifying wrappers (such as VecNormalize) ppo2 reports unnormalized episode rewards in `eqrewmean`. You can then log them in tensorboard format and view in tensorboard (via setting env variable `OPENAI_LOG_FORMAT=stdout,tensorboard`  or calling `logger.configure(format_strs=['stdout', 'tensorboard'])`. The way unnormalized rewards are obtained is by looking into info dict (which is populated by Monitor wrapper). 

As for whether VecNormalize should be used - I'd say the best bet is to try with and without and compare :) Original motivation in using VecNormalize was that mujoco observations are internal states of the mujoco physics simulator, and so components of the observation vector can vary in orders of magnitude. Roboschool observations are much more uniform in that regard; so normalization may not be needed. Which part of the ppo paper suggests that VecNormalize should be used? I don't think VecNormalize is mentioned anywhere in the paper explicitly.

**Comment by doviettung96** [comment]
Hi @pzhokhov ,
I have already tried unnormalized, state-normalized. Now I am trying state and reward normalization.
Basically, the results are really different from others. 
About the PPO paper, it is my mistake. But "GPU-Accelerated Robotic Simulation for Distributed Reinforcement Learning" does. The author whitens the observation. I suppose that he does not normalize the reward. As I have tried, the result is not quite close to the result in the paper.
